src/filters.rs

```diff
@@ -0,0 +1,1115 @@
+                } else {
+                    let updates = f.updates.clone();
+                    f.updates.clear();
+                    FilterChanges::Logs(updates)
+                }
+            }
+            FilterType::PendingTransaction(f) => {
+                if f.updates.is_empty() {
+                    FilterChanges::Empty(Default::default())
+                } else {
+                    let updates = f.updates.clone();
+                    f.updates.clear();
+                    FilterChanges::Hashes(updates)
+                }
+            }
+        };
+
+        Ok(changes)
+    }
+
+    /// Notify available filters of a newly produced block
+    pub fn notify_new_block(&mut self, hash: H256) {
+        self.filters.iter_mut().for_each(|(_, filter)| {
+            if let FilterType::Block(f) = filter {
+                f.updates.push(hash)
+            }
+        })
+    }
+
+    /// Notify available filters of a new pending transaction
+    pub fn notify_new_pending_transaction(&mut self, hash: H256) {
+        self.filters.iter_mut().for_each(|(_, filter)| {
+            if let FilterType::PendingTransaction(f) = filter {
+                f.updates.push(hash)
+            }
+        })
+    }
+
+    /// Notify available filters of a new transaction log
+    pub fn notify_new_log(&mut self, log: &Log, latest_block_number: U64) {
+        self.filters.iter_mut().for_each(|(_, filter)| {
+            if let FilterType::Log(f) = filter {
+                if f.matches(log, latest_block_number) {
+                    f.updates.push(log.clone());
+                }
+            }
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::testing::LogBuilder;
+
+    use super::*;
+
+    use maplit::{hashmap, hashset};
+
+    #[test]
+    fn test_add_block_filter() {
+        let mut filters = EthFilters::default();
+        let id = filters.add_block_filter().expect("failed adding filter");
+
+        assert_eq!(U256::from(1), id);
+        assert_eq!(
+            hashmap! {
+                U256::from(1) => FilterType::Block(BlockFilter { updates: vec![] })
+            },
+            filters.filters
+        );
+    }
+
+    #[test]
+    fn test_add_log_filter() {
+        let mut filters = EthFilters::default();
+        let id = filters
+            .add_log_filter(
+                BlockNumber::Latest,
+                BlockNumber::Number(U64::from(10)),
+                vec![H160::repeat_byte(0x1)],
+                [
+                    Some(hashset! { H256::repeat_byte(0x2) }),
+                    Some(hashset! { H256::repeat_byte(0x3), H256::repeat_byte(0x4) }),
+                    None,
+                    Some(hashset! {}),
+                ],
+            )
+            .expect("failed adding filter");
+
+        assert_eq!(U256::from(1), id);
+        assert_eq!(
+            hashmap! {
+                U256::from(1) => FilterType::Log(LogFilter {
```

Comment:
For some reason this is giving me an inline error in VSCode, but it compiles fine... any ideas?
<img width="731" alt="image" src="https://github.com/matter-labs/era-test-node/assets/1890113/d2ecfdc0-aeb8-4fde-891a-0f4a390289f2">

Reply:
That's not you, I missed boxing this in the test for the same reason it compiles - it is a test.

Reply:
TIL that Rust ignores test code in `.rs` files when compiling

Reply:
yeah, it needs the `--tests` with `cargo check` and `cargo build` to force it :)